### PR TITLE
Fix Collectors.toMap NPE bug

### PR DESCRIPTION
**Pull request by rozza**
Also only filter unprocessed items when calculating inserted results

JAVA-3515

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/async/client/AsyncMongoCollectionImpl.java [hunk 1]
```diff
@@ -1105,7 +1104,7 @@ private InsertOneResult toInsertOneResult(final com.mongodb.bulk.BulkWriteResult
     private InsertManyResult toInsertManyResult(final com.mongodb.bulk.BulkWriteResult result) {
         if (result.wasAcknowledged()) {
             return InsertManyResult.acknowledged(result.getInserts().stream()
-                    .collect(Collectors.toMap(BulkWriteInsert::getIndex, BulkWriteInsert::getId)));
+                    .collect(HashMap::new, (m, v) -> m.put(v.getIndex(), v.getId()), HashMap::putAll));
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/internal/async/client/AsyncMongoCollectionImpl.java]: This is the fix to the perf errors - caused by inserting multiple RawBsonDocuments - I'd only tested an insertOne.

File: driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java [hunk 2]
```diff
@@ -291,6 +291,9 @@ private BulkWriteResult getBulkWriteResult(final BsonDocument result) {
             if (!writeErrors.isEmpty()) {
                 writeRequests = writeRequests.filter(wr -> !writeErrors.contains(wr.getIndex()));
             }
+            if (payload.getPosition() < payload.size()) {
+                writeRequests = writeRequests.filter(wr -> wr.getIndex() <= payload.getPosition());
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java]: I noticed we weren't removing any unprocessed inserts.

File: driver-core/src/test/unit/com/mongodb/internal/connection/BulkWriteBatchCombinerSpecification.groovy [hunk 3]
```diff
@@ -47,13 +48,15 @@ class BulkWriteBatchCombinerSpecification extends Specification {
...
+        combiner.addResult(BulkWriteResult.acknowledged(INSERT, 1, 0, [], [new BulkWriteInsert(6, new BsonString('id1'))]))
+        combiner.addResult(BulkWriteResult.acknowledged(INSERT, 1, 0, [], [new BulkWriteInsert(3, new BsonString('id2'))]))
 
         when:
         def result = combiner.getResult()
 
         then:
-        result == BulkWriteResult.acknowledged(INSERT, 1, 0, [], [])
+        result == BulkWriteResult.acknowledged(INSERT, 2, 0, [],
+                [new BulkWriteInsert(3, new BsonString('id2')), new BulkWriteInsert(6, new BsonString('id1'))])
```
Comment (rozza) [inline driver-core/src/test/unit/com/mongodb/internal/connection/BulkWriteBatchCombinerSpecification.groovy]: This test just mirrors the upsert testing and ensures the ordering is as expected.

File: driver-sync/src/test/functional/com/mongodb/client/MongoCollectionTest.java [hunk 4]
```diff
@@ -172,6 +177,21 @@ public void testAggregationToACollection() {
         assertEquals(documents, result);
     }
 
+    @Test
+    public void bulkInsertRawBsonDocuments() {
```
Comment (rozza) [inline driver-sync/src/test/functional/com/mongodb/client/MongoCollectionTest.java]: Regression test.

**Comment by rozza** [comment]
https://evergreen.mongodb.com/version/5dd7bd791e2d1749ef50ab8a

File: driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java (hunk 2 above)
Comment (jyemin) [reply driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java]: Is there a test for this?  I can't tell if the one in the combiner spec is, but doesn't look like it.

File: driver-sync/src/test/functional/com/mongodb/client/MongoCollectionTest.java [hunk 5]
```diff
@@ -172,6 +177,21 @@ public void testAggregationToACollection() {
...
+        // given
+        List<RawBsonDocument> docs = asList(RawBsonDocument.parse("{a: 1}"), RawBsonDocument.parse("{a: 2}"));
+
+        // when
+        InsertManyResult result = collection.withDocumentClass(RawBsonDocument.class).insertMany(docs);
+
+        // then
+        Map<Integer, BsonValue> expectedResult = new HashMap<>();
+        expectedResult.put(0, null);
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/MongoCollectionTest.java]: I think we're fine as we are using BsonValues. So `null` means no value and `BsonNull.VALUE` means a null BsonValue.

File: driver-sync/src/test/functional/com/mongodb/client/MongoCollectionTest.java (hunk 4 above)
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/MongoCollectionTest.java]: Done - in the SmokeTest file

File: driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java (hunk 2 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/BulkWriteBatch.java]: Good call - added and fixed the bug in this line!

**Comment by rozza** [comment]
Updated based on feedback

File: driver-sync/src/test/functional/com/mongodb/client/MongoCollectionTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-sync/src/test/functional/com/mongodb/client/MongoCollectionTest.java]: Ah, ok.  Agreed

**Pull request closed without merging**